<h1 style="color:blue;"> Laboratorio 2 - Limpieza de Datos </h1>

In [265]:
   
    '''Empezamos importando las librerìas necesarias para realizar la limpieza de los datos existentes del Dataset '''

import pandas as pd
import numpy as np

In [266]:
 
    ''' Creamos la funciòn para importar nuestro Dataset '''

df= pd.read_csv('data_act_01.csv')

# Primer Error - Formato del Archivo

El primer error encontrado al ejecutar el dataset como se puede ver es que los datos se encuentran en una sola columna, esto se debe a que los datos se encuentran separados por un punto y coma(;) y no por una coma(,) como debería ser normalmente.

In [267]:
  
    ''' Enviamos a imprimir los datos existentes dentro del Dataset '''
    
df

CrimeId;OriginalCrimeTypeName;OffenseDate;CallTime;CallDateTime;Disposition;Address;City;State;AgencyId;Range;AddressType
0      160903280;Assault / Battery;2016-03-30T00:00:0...                                                                       
1      160912272;Homeless Complaint;2016-03-31T00:00:...                                                                       
2      160912590;Susp Info;2016-03-31T00:00:00;16:49;...                                                                       
3      160912801;Report;2016-03-31T00:00:00;17:38;201...                                                                       
4      160912811;594;2016-03-31T00:00:00;17:42;2016-0...                                                                       
...                                                  ...                                                                       
10046  160964210;Well Being Check;2016-04-05T00:00:00...                                                                       
10047  160964216;Traffic Stop;2016-04-05T00:00:00;23:...                                                                       
10048  160964227;Traffic Stop;2016-04-05T00:00:00;23:...                                                                       
10049  160964229;Passing Call;2016-04-05T00:00:00;23:...                                                                       
10050  160964249;Traffic Stop;2016-04-05T00:00:00;23:...                                                                       

[10051 rows x 1 columns]

# Corrección del Error

Para corregir el error presentado en el dataset se utilizó el parámetro sep donde se le especifíca el delimitador con el cual cuenta el archivo. Como se sabe este parámetro nos ayuda a separar los espacios que por defecto Python nos presenta. Y en este caso se le establece al separador que al encontrar un punto y coma en cada uno de los datos los separe, logrando de esta manera presentar los datos correctamente separados.

In [268]:
  
    ''' Creamos nuevamente la funciòn para importar nuestro Dataset agregando el parámetro sep, el cual nos permite separar los datos por espacios '''

df= pd.read_csv('data_act_01.csv', sep= ';')

In [269]:

    '''Se realiza una función mediante el uso del método head,el cual nos imprime los primeros datos del Dataset '''

df.head(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0  160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1  160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2  160912590             Susp Info  2016-03-31T00:00:00    16:49   
3  160912801                Report  2016-03-31T00:00:00    17:38   
4  160912811                   594  2016-03-31T00:00:00    17:42   
5  160913003                 Ref'd  2016-03-31T00:00:00    18:29   
6  160913050    Homeless Complaint  2016-03-31T00:00:00    18:43   
7  160913056    Homeless Complaint  2016-03-31T00:00:00    18:47   
8  160913078  Agg Assault / Adw Dv  2016-03-31T00:00:00    18:52   
9  160913103            Encampment  2016-03-31T00:00:00    18:57   

          CallDateTime Disposition                   Address           City  \
0  2016-03-30T18:42:00         REP   100 Block Of Chilton Av  San Francisco   
1  2016-03-31T15:31:00         GOA   2300 Block Of Market St  San Francisco   
2  2016-03-31T16:49:00         GOA   2300 Block Of Market St  San Francisco   
3  2016-03-31T17:38:00         GOA       500 Block Of 7th St  San Francisco   
4  2016-03-31T17:42:00         REP        Beale St/bryant St  San Francisco   
5  2016-03-31T18:29:00         GOA           16th St/pond St  San Francisco   
6  2016-03-31T18:43:00         ADV    Berwick Pl/harrison St  San Francisco   
7  2016-03-31T18:47:00         HAN    Florida St/mariposa St  San Francisco   
8  2016-03-31T18:52:00          ND  100 Block Of Genebern Wy  San Francisco   
9  2016-03-31T18:57:00         ADV   2700 Block Of Folsom St  San Francisco   

  State AgencyId  Range      AddressType  
0    CA        1    NaN  Premise Address  
1    CA        1    NaN  Premise Address  
2    CA        1    NaN  Premise Address  
3    CA        1    NaN  Premise Address  
4    CA        1    NaN     Intersection  
5    CA        1    NaN     Intersection  
6    CA        1    NaN     Intersection  
7    CA        1    NaN     Intersection  
8    CA        1    NaN  Premise Address  
9    CA        1    NaN  Premise Address

# Segundo Error - Valores Nulos

El segundo error encontrado dentro del dataset son valores nulos dentro de las columnas City, States y Range. Mediante el uso de la función isnull podemos observar el número exacto de valores faltantes dentro de las columnas.

In [270]:

    '''Mediante el uso del método isnull visualizamos el total de valores nulos existentes en cada una de las columnas '''

df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

# Corrección del Error

Para corregir el error encontrado se hace uso de la función fillna la cual reemplaza los valores faltantes o nulos por valores en cero, en el conjunto de datos existen valores nulos dentro de las columnas City, State y Range tal como se observa. Al implementar esta función podemos ver que donde existían valores NaN los reemplazo por valores en 0, ya que al contar con valores nulos nos puede generar muchos problemas a la hora de realizar un análisis de nuestros datos.

In [271]:

    '''Creamos una funciòn la cual nos ayudarà a reemplazar los valores vacios que existen en las columnas City, State y Range por valores en Cero, ya que al tener valores nulos nos puede generar muchos problemas a la hora de realizar un análisis de nuestros datos.  '''

df[['City', 'Range','State']] = df[['City', 'Range','State']].fillna(value=0)

Al verificar una vez más si existen o no datos nulos podemos ver que las columnas City, State y Range ya no presentar datos nulos.

In [272]:
  
    '''Enviamos a imprimir nuevamente los datos para verificar que se reemplazaron los valores nulos por valores en 0 '''

df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId  Range      AddressType  
0      San Francisco    CA        1    0.0  Premise Address  
1      San Francisco    CA        1    0.0  Premise Address  
2      San Francisco    CA        1    0.0  Premise Address  
3      San Francisco    CA        1    0.0  Premise Address  
4      San Francisco    CA        1    0.0     Intersection  
...              ...   ...      ...    ...              ...  
10046  San Francisco    CA        1    0.0     Intersection  
10047  San Francisco    CA        1    0.0     Intersection  
10048  San Francisco    CA        1    0.0     Intersection  
10049  San Francisco    CA        1    0.0     Geo-Override  
10050  San Francisco    CA        1    0.0     Intersection  

[10051 rows x 12 columns]

In [273]:

    '''Mediante el uso del método isnull visualizamos nuevamente el total de valores nulos, con el fin de verificar que ya no exista ningun valor nulo dentro de las columnas modificadas '''

df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
CallDateTime             0
Disposition              0
Address                  0
City                     0
State                    0
AgencyId                 0
Range                    0
AddressType              0
dtype: int64

# Tercer Error - Reemplazo de Datos Redundantes

El tercer error presentado fue el tener datos redundantes en las columnas City y State. En la columna City existian datos de la ciudad de San Francisco en mayúsculas y minúsculas, mientras que en la columna State se transforma CA a 1 para tener la columna totalmente numérica. 

# Columna City

In [274]:
    ''' Mediante la función groupby agrupamos todos los datos que existen dentro de la columna City '''

df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
SAN FRANCISCO        1                      1            1         1   
San Francisco     9665                   9665         9665      9665   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  Range  \
City                                                                        
0                       321          321      321    321       321    321   
 S                        1            1        1      1         1      1   
Brisbane                  1            1        1      1         1      1   
Daly City                 5            5        5      5         5      5   
Presidio                  3            3        3      3         3      3   
SAN FRANCISCO             1            1        1      1         1      1   
San Francisco          9665         9665     9665   9665      9665   9665   
Treasure Isla            51           51       51     51        51     51   
Yerba Buena               3            3        3      3         3      3   

               AddressType  
City                        
0                      321  
 S                       1  
Brisbane                 1  
Daly City                5  
Presidio                 3  
SAN FRANCISCO            1  
San Francisco         9665  
Treasure Isla           51  
Yerba Buena              3

# Corrección del Error

In [275]:
    ''' Mediante la función replace reemplazaremos las palabras que existan en mayúsculas por minúsculas '''

df['City']= df['City'].replace('SAN FRANCISCO','San Francisco')

In [276]:
    ''' Verificamos que el reemplazo de la fila se realizó correctamente '''

df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9666                   9666         9666      9666   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  Range  \
City                                                                        
0                       321          321      321    321       321    321   
 S                        1            1        1      1         1      1   
Brisbane                  1            1        1      1         1      1   
Daly City                 5            5        5      5         5      5   
Presidio                  3            3        3      3         3      3   
San Francisco          9666         9666     9666   9666      9666   9666   
Treasure Isla            51           51       51     51        51     51   
Yerba Buena               3            3        3      3         3      3   

               AddressType  
City                        
0                      321  
 S                       1  
Brisbane                 1  
Daly City                5  
Presidio                 3  
San Francisco         9666  
Treasure Isla           51  
Yerba Buena              3

In [277]:
    ''' Mediante la función replace reemplazaremos ahora la letra S por la ciudad de San Francisco '''

df['City']= df['City'].replace(' S','San Francisco')

In [278]:
    ''' Verificamos que el reemplazo de la fila se realizó correctamente '''

df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9667                   9667         9667      9667   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  Range  \
City                                                                        
0                       321          321      321    321       321    321   
Brisbane                  1            1        1      1         1      1   
Daly City                 5            5        5      5         5      5   
Presidio                  3            3        3      3         3      3   
San Francisco          9667         9667     9667   9667      9667   9667   
Treasure Isla            51           51       51     51        51     51   
Yerba Buena               3            3        3      3         3      3   

               AddressType  
City                        
0                      321  
Brisbane                 1  
Daly City                5  
Presidio                 3  
San Francisco         9667  
Treasure Isla           51  
Yerba Buena              3

In [279]:
    ''' Mediante la función replace reemplazaremos el valor de 0 por la ciudad de San Francisco '''

df['City']= df['City'].replace( 0,'San Francisco')

In [280]:
    ''' Verificamos que el reemplazo de la fila se realizó correctamente '''

df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9988                   9988         9988      9988   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  Range  \
City                                                                        
Brisbane                  1            1        1      1         1      1   
Daly City                 5            5        5      5         5      5   
Presidio                  3            3        3      3         3      3   
San Francisco          9988         9988     9988   9988      9988   9988   
Treasure Isla            51           51       51     51        51     51   
Yerba Buena               3            3        3      3         3      3   

               AddressType  
City                        
Brisbane                 1  
Daly City                5  
Presidio                 3  
San Francisco         9988  
Treasure Isla           51  
Yerba Buena              3

# Columna State

In [281]:
    ''' Mediante la función groupby agrupamos todos los datos que existen dentro de la columna State '''

df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  CallDateTime  \
State                                                                        
0            3                      3            3         3             3   
CA       10048                  10048        10048     10048         10048   

       Disposition  Address   City  AgencyId  Range  AddressType  
State                                                             
0                3        3      3         3      3            3  
CA           10048    10048  10048     10048  10048        10048

# Corrección del Error

In [282]:
    ''' Mediante la función replace reemplazaremos el valor de CA por el valor de 1 en la columna State '''

df['State']= df['State'].replace('CA','1')

In [283]:
    ''' Verificamos que el reemplazo de la fila se realizó correctamente '''

df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  CallDateTime  \
State                                                                        
0            3                      3            3         3             3   
1        10048                  10048        10048     10048         10048   

       Disposition  Address   City  AgencyId  Range  AddressType  
State                                                             
0                3        3      3         3      3            3  
1            10048    10048  10048     10048  10048        10048

# Corrección del Error

In [284]:
    '''Creamos una función mediante el uso del método to_numeric para modificar el formato del tipo de dato de la columna State '''

df['State']=pd.to_numeric(df['State'], errors='coerce')

In [285]:
df.dtypes

CrimeId                    int64
OriginalCrimeTypeName     object
OffenseDate               object
CallTime                  object
CallDateTime              object
Disposition               object
Address                   object
City                      object
State                      int64
AgencyId                  object
Range                    float64
AddressType               object
dtype: object

# Cuarto Error - Datos Duplicados

El cuarto error presentado es la existencia de datos duplicados dentro de la columna CrimeId. Mediante el uso de la función duplicated podemos verificar el número exacto de datos duplicados dentro de la columna CrimerId.

In [286]:

    '''Mediante el uso del método duplicated verficamos el número de datos duplicados dentro de la columna CrimeId   '''

df['CrimeId'].duplicated().sum()

4

In [287]:

    '''Creamos un filtro de los datos duplicados de la columna CrimeID y los transformamos en una lista.   '''

duplicates = df[df['CrimeId'].duplicated(keep=False)]['CrimeId'].tolist()

In [288]:

    '''Visualizamos los datos de los ID duplicados que existen dentro de la columna CrimeId.   '''

duplicates[:10]

[160913455, 160913455, 160913455, 160950496, 160950496, 160950496]

# Corrección del Error

Para corregir el error encontrado se crea un set de la lista de los valores, donde se le especifíca mediante el for que se debe recorrer el set de duplicados y debe comparar con cada uno de los datos en el Dataframe. Al encontrar un dato igual este debe sumarle 1 al número de CrimeId, de esta forma evitamos que se den valores duplicados dentro del CrimeId que es donde nos presenta el error.

In [289]:

    '''Realizamos una función la cual crea un set de la lista de los datos duplicados existentes en la columna CrimeId   '''

set_duplicates = set(duplicates)

In [290]:
'''Realizamos una función mediante el uso de un for el cual debe recorrer el set de duplicados y debe comparar con cada uno de los datos en el Dataframe. Al encontrar un dato igual este debe sumarle 1 al número de CrimeId, de esta forma evitamos que se den valores duplicados dentro del CrimeId que es donde nos presenta el error. '''
    
for r in set_duplicates:
    dup= df[df['CrimeId'] == r]
    df.loc[dup.iloc[1,:].name, 'CrimeId']= df.loc[dup.iloc[1,:].name]['CrimeId'] + 1    

Verificamos una vez más el número de duplicados dentro de la columna mediante la función duplicated y como podemos observar ya no presenta ningún dato duplicado.

In [291]:

    '''Visualizamos nuevamente los datos para verificar que no haya ningun registro duplicado dentro de la columna CrimeId '''

df.duplicated().sum()

0

# Quinto Error - Formato de Fechas y Horas

El quinto error presentado como se puede visualizar dento de las columnas OffenseDate y CallDateTime contienen la misma información y por ende el mismo formato de fecha y hora. Al visualizar que son datos 

In [292]:

    '''Mediante el uso del método loc seleccionamos las filas y las columnas de las columnas OffenseDate y CallDateTime para poder visualizar de mejor manera el tipo de formato que manejan en la Fecha del delito y la Fecha y hora de la llamada '''

df.loc[:, ['OffenseDate', 'CallDateTime']]

OffenseDate         CallDateTime
0      2016-03-30T00:00:00  2016-03-30T18:42:00
1      2016-03-31T00:00:00  2016-03-31T15:31:00
2      2016-03-31T00:00:00  2016-03-31T16:49:00
3      2016-03-31T00:00:00  2016-03-31T17:38:00
4      2016-03-31T00:00:00  2016-03-31T17:42:00
...                    ...                  ...
10046  2016-04-05T00:00:00  2016-04-05T23:38:00
10047  2016-04-05T00:00:00  2016-04-05T23:43:00
10048  2016-04-05T00:00:00  2016-04-05T23:48:00
10049  2016-04-05T00:00:00  2016-04-05T23:48:00
10050  2016-04-05T00:00:00  2016-04-05T23:54:00

[10051 rows x 2 columns]

# Corrección del Error

Para corregir el error encontrado creamos una función mediante el uso del método to_datetime. Este tipo de método nos permite cambiar el formato de la hora y fecha de las columnas OffenseDate, CallDateTime y CallTime.

In [293]:

    '''Creamos una función mediante el uso del método to_datetime para modificar el formato de hora y fecha correctamente dentro de la columna OffenseDate(Fecha del delito) '''

df['OffenseDate'] = pd.to_datetime(df['OffenseDate'])

In [294]:

    '''Creamos una función mediante el uso del método to_datetime para modificar el formato de hora y fecha correctamente dentro de la columna CallDateTime(Fecha y hora de la llamada ) '''

df['CallDateTime'] = pd.to_datetime(df['CallDateTime'])

In [295]:

    '''Creamos una función mediante el uso del método to_datetime para modificar el formato de hora correctamente dentro de la columna CallTime, la cual nos presenta el Tiempo de duración de llamada '''

df['CallTime']= pd.to_datetime(df['CallTime'], format='%H:%M:%S', errors='ignore')


In [296]:
''' Enviamos a imprimir nuevamente el DataFrame de nuestro dataset para verificar que el formato de las fechas y hora fueron modificadas correctamente '''

df

CrimeId OriginalCrimeTypeName OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30    18:42   
1      160912272    Homeless Complaint  2016-03-31    15:31   
2      160912590             Susp Info  2016-03-31    16:49   
3      160912801                Report  2016-03-31    17:38   
4      160912811                   594  2016-03-31    17:42   
...          ...                   ...         ...      ...   
10046  160964210      Well Being Check  2016-04-05    23:38   
10047  160964216          Traffic Stop  2016-04-05    23:43   
10048  160964227          Traffic Stop  2016-04-05    23:48   
10049  160964229          Passing Call  2016-04-05    23:48   
10050  160964249          Traffic Stop  2016-04-05    23:54   

             CallDateTime   Disposition                  Address  \
0     2016-03-30 18:42:00           REP  100 Block Of Chilton Av   
1     2016-03-31 15:31:00           GOA  2300 Block Of Market St   
2     2016-03-31 16:49:00           GOA  2300 Block Of Market St   
3     2016-03-31 17:38:00           GOA      500 Block Of 7th St   
4     2016-03-31 17:42:00           REP       Beale St/bryant St   
...                   ...           ...                      ...   
10046 2016-04-05 23:38:00           GOA    Hollister Av/hawes St   
10047 2016-04-05 23:43:00           ADV   California St/mason St   
10048 2016-04-05 23:48:00           CIT   Rosella Ct/onondaga Av   
10049 2016-04-05 23:48:00  Not recorded          0 Blk Herbst Rd   
10050 2016-04-05 23:54:00           ADV        Grant Av/geary St   

                City  State AgencyId  Range      AddressType  
0      San Francisco      1        1    0.0  Premise Address  
1      San Francisco      1        1    0.0  Premise Address  
2      San Francisco      1        1    0.0  Premise Address  
3      San Francisco      1        1    0.0  Premise Address  
4      San Francisco      1        1    0.0     Intersection  
...              ...    ...      ...    ...              ...  
10046  San Francisco      1        1    0.0     Intersection  
10047  San Francisco      1        1    0.0     Intersection  
10048  San Francisco      1        1    0.0     Intersection  
10049  San Francisco      1        1    0.0     Geo-Override  
10050  San Francisco      1        1    0.0     Intersection  

[10051 rows x 12 columns]

In [297]:

    '''Mediante el uso de la función dtypes podemos visulizar que el tipo de dato de las columnas OffenseDate y CallDateTime cambiaron al formato datetime '''

df.dtypes

CrimeId                           int64
OriginalCrimeTypeName            object
OffenseDate              datetime64[ns]
CallTime                         object
CallDateTime             datetime64[ns]
Disposition                      object
Address                          object
City                             object
State                             int64
AgencyId                         object
Range                           float64
AddressType                      object
dtype: object

# Sexto Error - Eliminación de Columnas Innecesarias

El sexto error que se presentó fue que dentro de la columna Range no contaba con ningún tipo de dato, a lo que se entiende que esta columna no representa nada dentro del Dataset. Cabe recalcar que al realizar un shape visualizamos que existen 10051 datos y es el mismo número de valores nulos que presentaba la columna mencionada antes de realizar el reemplazo de datos nulos. Es por esto que se decidió eliminar por completo la columna.


In [298]:

    '''Mediante el uso del método shape visualizamos el valor total de los datos y las columnas con las que cuenta el Dataset  '''

df.shape

(10051, 12)

In [299]:

    '''Enviamos a imprimir los datos del Dataset para verificar los datos nulos que existen dentro de la columna Range'''

df

CrimeId OriginalCrimeTypeName OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30    18:42   
1      160912272    Homeless Complaint  2016-03-31    15:31   
2      160912590             Susp Info  2016-03-31    16:49   
3      160912801                Report  2016-03-31    17:38   
4      160912811                   594  2016-03-31    17:42   
...          ...                   ...         ...      ...   
10046  160964210      Well Being Check  2016-04-05    23:38   
10047  160964216          Traffic Stop  2016-04-05    23:43   
10048  160964227          Traffic Stop  2016-04-05    23:48   
10049  160964229          Passing Call  2016-04-05    23:48   
10050  160964249          Traffic Stop  2016-04-05    23:54   

             CallDateTime   Disposition                  Address  \
0     2016-03-30 18:42:00           REP  100 Block Of Chilton Av   
1     2016-03-31 15:31:00           GOA  2300 Block Of Market St   
2     2016-03-31 16:49:00           GOA  2300 Block Of Market St   
3     2016-03-31 17:38:00           GOA      500 Block Of 7th St   
4     2016-03-31 17:42:00           REP       Beale St/bryant St   
...                   ...           ...                      ...   
10046 2016-04-05 23:38:00           GOA    Hollister Av/hawes St   
10047 2016-04-05 23:43:00           ADV   California St/mason St   
10048 2016-04-05 23:48:00           CIT   Rosella Ct/onondaga Av   
10049 2016-04-05 23:48:00  Not recorded          0 Blk Herbst Rd   
10050 2016-04-05 23:54:00           ADV        Grant Av/geary St   

                City  State AgencyId  Range      AddressType  
0      San Francisco      1        1    0.0  Premise Address  
1      San Francisco      1        1    0.0  Premise Address  
2      San Francisco      1        1    0.0  Premise Address  
3      San Francisco      1        1    0.0  Premise Address  
4      San Francisco      1        1    0.0     Intersection  
...              ...    ...      ...    ...              ...  
10046  San Francisco      1        1    0.0     Intersection  
10047  San Francisco      1        1    0.0     Intersection  
10048  San Francisco      1        1    0.0     Intersection  
10049  San Francisco      1        1    0.0     Geo-Override  
10050  San Francisco      1        1    0.0     Intersection  

[10051 rows x 12 columns]

# Corrección del Error

In [300]:

    '''Mediante el uso del método drop eliminamos la columna Range, ya que al no tener ningún tipo de dato no es necesaria la limpieza de los datos '''

df.drop(['Range'], axis='columns', inplace=True)

In [301]:

    '''Visualizamos nuevamente los datos para verificar que se eliminó la columna Range correctamente '''

df

CrimeId OriginalCrimeTypeName OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30    18:42   
1      160912272    Homeless Complaint  2016-03-31    15:31   
2      160912590             Susp Info  2016-03-31    16:49   
3      160912801                Report  2016-03-31    17:38   
4      160912811                   594  2016-03-31    17:42   
...          ...                   ...         ...      ...   
10046  160964210      Well Being Check  2016-04-05    23:38   
10047  160964216          Traffic Stop  2016-04-05    23:43   
10048  160964227          Traffic Stop  2016-04-05    23:48   
10049  160964229          Passing Call  2016-04-05    23:48   
10050  160964249          Traffic Stop  2016-04-05    23:54   

             CallDateTime   Disposition                  Address  \
0     2016-03-30 18:42:00           REP  100 Block Of Chilton Av   
1     2016-03-31 15:31:00           GOA  2300 Block Of Market St   
2     2016-03-31 16:49:00           GOA  2300 Block Of Market St   
3     2016-03-31 17:38:00           GOA      500 Block Of 7th St   
4     2016-03-31 17:42:00           REP       Beale St/bryant St   
...                   ...           ...                      ...   
10046 2016-04-05 23:38:00           GOA    Hollister Av/hawes St   
10047 2016-04-05 23:43:00           ADV   California St/mason St   
10048 2016-04-05 23:48:00           CIT   Rosella Ct/onondaga Av   
10049 2016-04-05 23:48:00  Not recorded          0 Blk Herbst Rd   
10050 2016-04-05 23:54:00           ADV        Grant Av/geary St   

                City  State AgencyId      AddressType  
0      San Francisco      1        1  Premise Address  
1      San Francisco      1        1  Premise Address  
2      San Francisco      1        1  Premise Address  
3      San Francisco      1        1  Premise Address  
4      San Francisco      1        1     Intersection  
...              ...    ...      ...              ...  
10046  San Francisco      1        1     Intersection  
10047  San Francisco      1        1     Intersection  
10048  San Francisco      1        1     Intersection  
10049  San Francisco      1        1     Geo-Override  
10050  San Francisco      1        1     Intersection  

[10051 rows x 11 columns]

In [302]:
    '''Mediante el uso de la función drop eliminamos la columna CallDateTime ya que esta columna muestras los mismos datos de la columna  CallDateTime '''

df.drop(['CallDateTime'], axis='columns', inplace=True)

In [303]:
  '''Visualizamos nuevamente los datos para verificar que se eliminó la columna CallDateTime correctamente '''

df

CrimeId OriginalCrimeTypeName OffenseDate CallTime   Disposition  \
0      160903280     Assault / Battery  2016-03-30    18:42           REP   
1      160912272    Homeless Complaint  2016-03-31    15:31           GOA   
2      160912590             Susp Info  2016-03-31    16:49           GOA   
3      160912801                Report  2016-03-31    17:38           GOA   
4      160912811                   594  2016-03-31    17:42           REP   
...          ...                   ...         ...      ...           ...   
10046  160964210      Well Being Check  2016-04-05    23:38           GOA   
10047  160964216          Traffic Stop  2016-04-05    23:43           ADV   
10048  160964227          Traffic Stop  2016-04-05    23:48           CIT   
10049  160964229          Passing Call  2016-04-05    23:48  Not recorded   
10050  160964249          Traffic Stop  2016-04-05    23:54           ADV   

                       Address           City  State AgencyId      AddressType  
0      100 Block Of Chilton Av  San Francisco      1        1  Premise Address  
1      2300 Block Of Market St  San Francisco      1        1  Premise Address  
2      2300 Block Of Market St  San Francisco      1        1  Premise Address  
3          500 Block Of 7th St  San Francisco      1        1  Premise Address  
4           Beale St/bryant St  San Francisco      1        1     Intersection  
...                        ...            ...    ...      ...              ...  
10046    Hollister Av/hawes St  San Francisco      1        1     Intersection  
10047   California St/mason St  San Francisco      1        1     Intersection  
10048   Rosella Ct/onondaga Av  San Francisco      1        1     Intersection  
10049          0 Blk Herbst Rd  San Francisco      1        1     Geo-Override  
10050        Grant Av/geary St  San Francisco      1        1     Intersection  

[10051 rows x 10 columns]

# Archivo Json

Una vez realizada la limpieza de los datos del Dataset seleccionado se procede a crear un documento JSON, el cual contiene toda la información de las columnas y filas modificadas.

In [304]:

    '''Creamos una función mediante el uso del método to_json, el cual nos permite crear  convertir un DataFrame de pandas en una lista de diccionarios de los datos limpios del Dataset'''

df.to_json('data_act_01_Limpiada.json', orient='records', lines=True)